workspaces/adventure-pack/goodies/typescript/Map.groupBy/index.ts

```diff
@@ -0,0 +1,29 @@
+declare global {
+  interface MapConstructor {
+    groupBy<K, V>(
+      iterable: Iterable<V>,
+      callbackFn: (value: V, index: number) => K,
+    ): Map<K, V[]>;
+  }
+}
+
+Map.groupBy = function <K, V>(
+  iterable: Iterable<V>,
+  callbackFn: (value: V, index: number) => K,
+): Map<K, V[]> {
+  const groups = new Map<K, V[]>();
+  let index = 0;
+  for (const value of iterable) {
+    const key = callbackFn(value, index);
+    if (!groups.has(key)) {
+      groups.set(key, []);
+    }
+    groups.get(key)!.push(value);
```

Comment:
When it's not too annoying, I've been trying to avoid `!` and other ostensibly unsafe TypeScript features in the goody code. Although it would be safe in this case, avoiding it is not too bad:

```ts
const group = groups.get(key);
if (group == null) {
  groups.set(key, [value]);
} else {
  group.push(value);
}
```

As a marginal potential benefit, this is one less map lookup.

Also, TypeScript would benefit from having something like Java's [`computeIfAbsent`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Map.html#computeIfAbsent(K,java.util.function.Function)), so I think we found our next goody idea 😄

Reply:
Thanks for pointing me in the direction of why using the bang post-fix expression operator is probably not a good habit to be in! Definitely learned something 😃 

P.S. - If you're suggesting I take on a `computeIfAbsent` TS port then I can definitely do that! Not sure what level of autonomy you're looking for from contributors but figured it couldn't hurt to clarify this time at least.

Reply:
Since folks are contributing on a volunteer basis in their free time, I have no particular expectations -- I'm grateful for whichever way folks would choose to contribute. That said, autonomy and initiative are highly appreciated from anyone who wants to engage with the project more -- please feel free to propose new goody ideas, as well as any other kinds of improvements. Issues or the Discord channel are a great way to do that.

And if you'd like to implement any of the ideas too, that'd be even more amazing. Thanks for asking and for contributing!

Reply:
(For this particular PR, I don't think it should block on adding a new goody, but we could come back and use that goody here later on.)

Reply:
Is there an update to this file that I'm not seeing? I think this is the only thing keeping this PR from getting merged.

Reply:
Yes, there was an update you were not seeing.... because I had not committed it 🤕 

Thanks for all the help on this one!